### DISCUSSION TITLE
Which functions run first?

### DISCUSSION THREAD

**Discussion by tyzion** [apostrophecms/sanitize-html#617] [General]
Hi, I have a question.

Is there an order in the execution of the functions I can use in sanitize options?

E.g. if I use both **transformTags** and **exclusiveFilter**, which will run first?
I've tried using console.log to understand, but couldn't

thanks

**Comment by BoDonkey** [comment]
In a quick test (not through looking at the code carefully) it looks to me like any "built-in" functions, like 'transformTags' is going to run before any custom functions. I didn't dig enough to see the order if you supply multiple custom functions. If this isn't optimal, you could think of combining all of the changes you need into a single custom function. Not sure of your use-case.

**Reply by tyzion** [reply]
what do you mean by custom function?
Anyway, by console.log, I noticed that **exclusiveFilter** starts first. But when it reaches a tag that's in common with **transformTags** they both will run

**Reply by BoDonkey** [reply]
Sorry - bad phrasing, and now that I think about it, bad test - DOH! By 'built-in' I meant a use case like 
```
transformTags: {
    'ol': 'ul',
  }
```
As opposed to one where you supply a full function, like
```
textFilter: function(text, tagName)
...
```
Can you share what you are trying to accomplish?

**Reply by tyzion** [reply]
Sure. But it was more of a "how does it work question"

What I want to do is ignore some tags in given situations, and do some filtering transformation in other.
I'm doing that without no problem.
Specifically what I wanted to know was if **transformTags** would run after I've filtered the tags with my **exclusiveFilter**, since it would be useless to filter something I want to delete.

anyway, here's my code:
```
sanitize(text, {
    transformTags: {
        'a': sanitize.simpleTransform('a', { target: '_blank' }),
        'img': function(tagName, attribs) {
            if (...) {
                return { tagName: 'p', text: '[image]' }
            }
        }
    },
    exclusiveFilter: function(frame) {
        if (frame.tag === 'div' && frame.attribs.id) {
            if(frame.attribs.id.includes('...')) return true
        }
        return excludeFrame(provider, frame)
    }
})
```

**Comment by BoDonkey** [comment]
Just taking a quick look at the code. TransformTags is fired by the htmlparser2 `onopentag` handler, while exclusiveFilter is fired by the `onclosetag` handler. So I guess TransformTags would probably run first.

**Reply by tyzion** [reply]
allright! thank you very much =-)
